<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>tcR: An R package for TCR repertoire advanced data analysis</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>tcR</h1>

        <h4>An R package for T-cell receptor repertoire advanced data analysis </h4>

        <p>T-cell receptor, TCR repertoire, TCR analysis, immunoinformatics, computational immunology</p>

        <p><a href="http://imminfo.github.io/tcr/">Paper in <i>Journal of Something</i> (stay tuned...) </a></p>

        <ul>
          <li><a href="tcrvignette.pdf">View PDF <strong>Vignette</strong></a></li>
          <li><a href="https://github.com/imminfo/tcr/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/imminfo/tcr">View On <strong>GitHub</strong></a></li>
        </ul>
        
        <p class="view"><a href="http://cran.r-project.org/web/packages/tcR/index.html">tcR at CRAN <small>Package info at CRAN</small></a></p>
        
        <p class="view"><a href="http://imminfo.github.io">Immunoinformatics Page <small>Our other projects and tools.</small></a></p>

      </header>
      <section>
        <h3>

<h3>
<a name="overview" class="anchor" href="#overview"><span class="octicon octicon-link"></span></a>Overview</h3>

<p><strong>tcR</strong> is a platform designed for TCR repertoire data analysis in R after preprocessing data with <a href="http://www.nature.com/nmeth/journal/v10/n9/full/nmeth.2555.html">MiTCR</a> or other CDR3 extraction and gene alignment software tools. With the power and flexibility of R language and
procedures supported by <strong>tcR</strong> users can perform advanced statistical analysis of TCR repertoires. Features of the package:</p>

<ul>
  <li>Parsers for outputs of various tools for CDR3 extraction and gene segments alignment</li>
  <i>(currently only MiTCR implemented)</i>
  <li>Data manipulation</li>
  <i>(in-frame / out-of-frame sequences subsetting, clonotype motif search)</i>
  <li>Descriptive statistics</li>
  <i>(number of reads, number of clonotypes)</i>
  <li>Shared clones statistics<br>
  <i>(number of shared clones / clonotypes, using V-segments or not; sequential intersection among the most abundant clones ("top-cross"))</i></li>
  <li>Repertoire comparison</li>
  <i>(Jaccard index, Morisita's overlap index, Horn's index, Tversky index, overlap coefficient)</i>
  <li>V- and J-segments usage and it's analysis<br>
  <i>(PCA, Shannon Entropy, Jensen-Shannon Divergence)</i></li>
  <li>Diversity evaluation<br>
  <i>(ecological diversity index, Gini index, inverse Simpson index, rarefaction analysis)</i></li>
  <li>Artificial repertoire generation (beta chain only, for now)</li>
  <li>Spectratyping</li>
  <li>Various visualisation procedures</li>
  <li>Mutation networks<br>
  <i>(graphs, in which vertices represent CDR3 nucleotide / amino acid sequences and edges are connecting similar sequences with low hamming or edit distance between them)</i></li>
</ul>

<p>Manual: <a href="tcrvignette.pdf">tcR package vignette</a></p>

<h3>
<a name="binaries" class="anchor" href="#binaries"><span class="octicon octicon-link"></span></a>Installation</h3>
<p>The package is available through CRAN and through <i>devtools::install_github</i>.
<p>CRAN version: <strong>1.1</strong></p>
<p>Dev version: <strong>1.3</strong></p>
<p>Install the latest release version from CRAN with:</p>
<pre><code>install.packages("tcR")
</code></pre>
<p>Install the latest development version <strong>(strongly recommended)</strong> with:</p>
<pre><code># install.packages("devtools")
devtools::install_github("imminfo/tcR", build_vignettes = FALSE)
</code></pre>

<h3>
<a name="quickstart" class="anchor" href="#quickstart"><span class="octicon octicon-link"></span></a>Quick Start</h3>

<p>The package provides two <a href="http://rmarkdown.rstudio.com">RMarkdown</a> documents with exploratory analysis of a single or a groups of repertoires and crossanalysis of a group of repertoires..
  User just need to change paths in this documents to their <a href = "http://www.nature.com/nmeth/journal/v10/n9/full/nmeth.2555.html">MiTCR</a> data and have the <a href = "http://yihui.name/knitr/">knitr</a> package installed.
</p>
<p>
  Exploratory analysis (<a href = "twb12.lib.report.html">example for two repertoires</a>) stored at <i>"tcR-package-path/inst/library.report.Rmd"</i>
  and included statistics and visualisation of number of clones, clonotypes, in- and out-of-frames, unique amino acid CDR3
  sequences, V- and J-usage, most frequent k-mers, rarefaction analysis.
</p>
<p>
  Analysis of a group of repertoires (<a href = "twb.cross.report.html">example for four repertoires</a>) stored at <i>"tcR-package-path/inst/crossanalysis.report.Rmd"</i>
  and included statistics and visualisation of number of shared clones and clonotypes, V-usage for individuals and groups, J-usage for individuals,
  Jensen-Shannon divergence among V-usages of repertoires and top-cross.
</p>

<h3>
<a name="examples" class="anchor" href="#examples"><span class="octicon octicon-link"></span></a>Examples</h3>

<p>First, load the list with data frames with beta chain data of 2 twin pairs and V- and J-genes alphabets for human subjects:</p>
<pre><code>library(tcR)
data(twb)
</code></pre>

<p>Get numbers of clones, in-frames and out-of-frames, unique amino acid CDR3 sequences and other basic statistics:</p>
<pre><code>mitcr.stats(twb)
</code></pre>

<p>Compute V-usage of the first subject:</p>
<pre><code>freq.Vb(twb[[1]])
</code></pre>

<p>Compute V-usage of all subjects and visualise it through a histogram:</p>
<pre><code>vis.V.usage(twb, .dodge = T)
</code></pre>

<p>Compute a pairwise number of shared clones and plot a heatmap of them:</p>
<pre><code>twb.int <- intersect(twb, 'n0e')
vis.heatmap(twb.int)
</code></pre>

<p>Perform a rarefaction analysis and plot the resulting rarefaction curve on a log x-axis:</p>
<pre><code>twb.rar <- rarefaction(twb, .col = "Read.count")
vis.rarefaction(twb, .log = T)
</code></pre>

<p>For more examples see the <a href="tcrvignette.pdf">vignette</a></p>.

<h3>
<a name="authors-and-contributors" class="anchor" href="#authors-and-contributors"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>The package was developed by members of <a href = "http://labcfg.ibch.ru/tcr">Laboratory of Functional and Comparative Genomics, Moscow, Russia</a> and
<a href = "http://milaboratory.com/">Laboratory of Genomics of Adaptive Immunity, Moscow, Russia</a>.
  Main developers are Vadim Nazarov <a href="https://github.com/vadimnazarov" class="user-mention">@vadimnazarov</a> and Mikhail Pogorelyy <a href="https://github.com/pogorely" class="user-mention">@pogorely</a>.
</p>

<h3>
<a name="support-or-contact" class="anchor" href="#support-or-contact"><span class="octicon octicon-link"></span></a>Contact</h3>

<p>For any question contact <a href="https://github.com/vadimnazarov" class="user-mention">@vadimnazarov</a> or raise an issue at <a href="https://github.com/imminfo/tcr/issues">Issues</a>.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/vadimnazarov">Vadim N. (@vadimnazarov)</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>

  </body>
</html>
